### ISSUE TITLE
test_aniso_pair.py in Debug builds

### ISSUE THREAD

**Issue by joaander** [glotzerlab/hoomd-blue#1282]
## Description

<!-- Describe the problem. -->

## Script

```bash
$ cmake ~/devel/hoomd -DCMAKE_BUILD_TYPE=Debug -GNinja
$ ninja
$ python3 -m pytest hoomd/md/pytest/test_aniso_pair.py -v -x -ra
```

<!-- Attach any input files needed to execute the script. -->

## Output

<!-- What output did you get? -->
```
==================================================================== test session starts ====================================================================
platform linux -- Python 3.9.5, pytest-7.1.0, pluggy-1.0.0 -- /home/joaander/venv/bin/python3                                                                
cachedir: .pytest_cache                                                                                                                                      
rootdir: /home/joaander/build/hoomd                                                                                                                          
plugins: anyio-3.5.0                                                                                                                                         
collected 38 items                                                                                                                                          
                                                                                                                                                             
hoomd/md/pytest/test_aniso_pair.py::test_mode[CPU-mode0] PASSED                                                                                       [  2%]
hoomd/md/pytest/test_aniso_pair.py::test_mode[CPU-mode1] PASSED                                                                                       [  5%]
hoomd/md/pytest/test_aniso_pair.py::test_rcut[CPU-2.5] PASSED                                                                                         [  7%]
hoomd/md/pytest/test_aniso_pair.py::test_rcut[CPU-4.0] PASSED                                                                                         [ 10%]
hoomd/md/pytest/test_aniso_pair.py::test_rcut[CPU-1.0] PASSED                                                                                         [ 13%]
hoomd/md/pytest/test_aniso_pair.py::test_rcut[CPU-0.01] PASSED                                                                                        [ 15%]
hoomd/md/pytest/test_aniso_pair.py::test_setting_params_and_shape[CPU-Dipole-1] PASSED                                                                [ 18%]
hoomd/md/pytest/test_aniso_pair.py::test_setting_params_and_shape[CPU-GayBerne-1] PASSED                                                              [ 21%]
hoomd/md/pytest/test_aniso_pair.py::test_setting_params_and_shape[CPU-ALJ-1] PASSED                                                                   [ 23%]
hoomd/md/pytest/test_aniso_pair.py::test_setting_params_and_shape[CPU-ALJ-2] PASSED                                                                   [ 26%]
hoomd/md/pytest/test_aniso_pair.py::test_run[CPU-Dipole-1] PASSED                                                                                     [ 28%]
hoomd/md/pytest/test_aniso_pair.py::test_run[CPU-GayBerne-1] PASSED                                                                                   [ 31%]
hoomd/md/pytest/test_aniso_pair.py::test_run[CPU-ALJ-1] PASSED                                                                                        [ 34%]
hoomd/md/pytest/test_aniso_pair.py::test_run[CPU-ALJ-2] PASSED                                                                                        [ 36%]
hoomd/md/pytest/test_aniso_pair.py::test_aniso_force_computes[CPU-ALJ0] PASSED                                                                        [ 39%] 
hoomd/md/pytest/test_aniso_pair.py::test_aniso_force_computes[CPU-ALJ1] PASSED                                                                        [ 42%]
hoomd/md/pytest/test_aniso_pair.py::test_aniso_force_computes[CPU-ALJ2] Fatal Python error: Aborted
                                                                                                                                                             
Current thread 0x00007f1e1fb5a740 (most recent call first):                                                                                                  
  File "/home/joaander/build/hoomd/hoomd/md/force.py", line 84 in energies                                                                                   
  File "/home/joaander/build/hoomd/hoomd/logging.py", line 438 in wrapped_func                 
  File "/home/joaander/build/hoomd/hoomd/md/pytest/test_aniso_pair.py", line 456 in test_aniso_force_computes
  File "/home/joaander/venv/lib/python3.9/site-packages/_pytest/python.py", line 192 in pytest_pyfunc_call   
  File "/home/joaander/venv/lib/python3.9/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/joaander/venv/lib/python3.9/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/joaander/venv/lib/python3.9/site-packages/pluggy/_hooks.py", line 265 in __call__                                                             
  File "/home/joaander/venv/lib/python3.9/site-packages/_pytest/python.py", line 1761 in runtest                                                            
  File "/home/joaander/venv/lib/python3.9/site-packages/_pytest/runner.py", line 166 in pytest_runtest_call                                                 
  File "/home/joaander/venv/lib/python3.9/site-packages/pluggy/_callers.py", line 39 in _multicall                                                           
  File "/home/joaander/venv/lib/python3.9/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/joaander/venv/lib/python3.9/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/joaander/venv/lib/python3.9/site-packages/_pytest/runner.py", line 259 in <lambda>
  File "/home/joaander/venv/lib/python3.9/site-packages/_pytest/runner.py", line 338 in from_call
  File "/home/joaander/venv/lib/python3.9/site-packages/_pytest/runner.py", line 258 in call_runtest_hook                                                   
  File "/home/joaander/venv/lib/python3.9/site-packages/_pytest/runner.py", line 219 in call_and_report
  File "/home/joaander/venv/lib/python3.9/site-packages/_pytest/runner.py", line 130 in runtestprotocol                                                     
  File "/home/joaander/venv/lib/python3.9/site-packages/_pytest/runner.py", line 111 in pytest_runtest_protocol                                             
  File "/home/joaander/venv/lib/python3.9/site-packages/pluggy/_callers.py", line 39 in _multicall                                                          
  File "/home/joaander/venv/lib/python3.9/site-packages/pluggy/_manager.py", line 80 in _hookexec                                                           
  File "/home/joaander/venv/lib/python3.9/site-packages/pluggy/_hooks.py", line 265 in __call__                                                             
  File "/home/joaander/venv/lib/python3.9/site-packages/_pytest/main.py", line 347 in pytest_runtestloop                                                    
  File "/home/joaander/venv/lib/python3.9/site-packages/pluggy/_callers.py", line 39 in _multicall                                                          
  File "/home/joaander/venv/lib/python3.9/site-packages/pluggy/_manager.py", line 80 in _hookexec                                                           
  File "/home/joaander/venv/lib/python3.9/site-packages/pluggy/_hooks.py", line 265 in __call__                                                             
  File "/home/joaander/venv/lib/python3.9/site-packages/_pytest/main.py", line 322 in _main                                                                 
  File "/home/joaander/venv/lib/python3.9/site-packages/_pytest/main.py", line 268 in wrap_session                                                          
  File "/home/joaander/venv/lib/python3.9/site-packages/_pytest/main.py", line 315 in pytest_cmdline_main                                                   
  File "/home/joaander/venv/lib/python3.9/site-packages/pluggy/_callers.py", line 39 in _multicall                                                          
  File "/home/joaander/venv/lib/python3.9/site-packages/pluggy/_manager.py", line 80 in _hookexec                                                           
  File "/home/joaander/venv/lib/python3.9/site-packages/pluggy/_hooks.py", line 265 in __call__                                                             
  File "/home/joaander/venv/lib/python3.9/site-packages/_pytest/config/__init__.py", line 164 in main                                                       
  File "/home/joaander/venv/lib/python3.9/site-packages/_pytest/config/__init__.py", line 187 in console_main
  File "/home/joaander/venv/lib/python3.9/site-packages/pytest/__main__.py", line 5 in <module>
  File "/usr/lib/python3.9/runpy.py", line 87 in _run_code 
  File "/usr/lib/python3.9/runpy.py", line 197 in _run_module_as_main     
zsh: abort (core dumped)  python3 -m pytest hoomd/md/pytest/test_aniso_pair.py -v -x -ra
```

`gdb` reports this assert is failing:
```
#4  0x00007fffb8a589c4 in hoomd::md::EvaluatorPairALJ<3u>::evaluate (this=0x7fffffff81e0, force=..., pair_eng=@0x7fffffff7e08: 0, energy_shift=false, 
    torque_i=..., torque_j=...) at /home/joaander/devel/hoomd/hoomd/md/EvaluatorPairALJ.h:584
584                     assert(success && !overlap);
```

## Expected output

<!-- What output did you expect? -->
All tests pass, as they do in Release builds.

## Configuration

<!-- What is your system configuration? -->

<!-- Remove items that do not apply. -->

Platform:
- Mac
- Linux
- CPU

Installation method:
- Compiled from source

<!-- What software versions do you have? -->

Versions:
- Python version: 3.9.5
- **HOOMD-blue** version: `trunk-patch`

## Developer

<!-- Who should implement the fix? We would welcome your contribution! -->
@tommy-waltmann, please investigate. @vyasr, your input would be appreciated if you have time.

**Comment by tommy-waltmann** [comment]
I've seen this before in debug builds when I was converting to the v3 API. I chatted with Thi and @vyasr about it offline, but I don't remember the exact reason they said this can happen.

**Comment by vyasr** [comment]
I'm in the middle of a busy patch and probably won't be able to investigate myself, but I can help Tommy with his investigations. 

This failure is typically caused by attempting to simulate a shape for which GJK does not converge well enough. Usually the problem manifests as an accumulation of errors in the solution over multiple time steps leading to invalid energies and forces, which in turn eventually leads to unphysical collisions that end in overlapping configurations. It's possible that with sufficient thought we could come up with a more stable algorithm, but GJK is a sufficiently active area of research in other communities that improving it would be a significant research effort in and of itself. I don't know of a better alternative here, though. I left the assertion there as a way to help users debug simulations where they observe unphysical behavior because I didn't see any nicer way to bubble that up to the user in Python in HOOMD 2.x. Now that more of the simulation loop logic is handled in Python we could make it possible for runs to set some sort of error code in C++ to indicate that something went wrong, then the Python could inspect that and raise an Exception?

**Comment by joaander** [comment]
This test is just a single force computation of a given configuration of particles. It is possible that configuration is overlapping or bad.

If needed, one could add a single "error" flag that PotentialPair or PotentialPairAniso sets which the caller uses to raise an exception.

**Comment by vyasr** [comment]
Resolved in #1292.
